pkg/kubefnord/unjoin.go

```diff
@@ -0,0 +1,285 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubefnord
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	fedclient "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	controllerutil "github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/kubefnord/options"
+	"github.com/kubernetes-sigs/federation-v2/pkg/kubefnord/util"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	crclient "k8s.io/cluster-registry/pkg/client/clientset/versioned"
+)
+
+var (
+	unjoin_long = `
+		Unjoin removes a cluster from a federation.
+		Current context is assumed to be a Kubernetes cluster
+		with an aggregated federation API server. Please use
+		the --host-cluster-context flag otherwise.`
+	unjoin_example = `
+		# Unjoin a cluster from a federation by specifying the
+		# cluster name and the context name of the federation
+		# control plane's host cluster. Cluster name must be
+		# a valid RFC 1123 subdomain name. Cluster context
+		# must be specified if the cluster name is different
+		# than the cluster's context in the local kubeconfig.
+		kubefnord unjoin foo --host-cluster-context=bar`
+)
+
+type unjoinFederation struct {
+	options.SubcommandOptions
+	unjoinFederationOptions
+}
+
+type unjoinFederationOptions struct {
+	clusterContext     string
+	removeFromRegistry bool
+}
+
+// Bind adds the unjoin specific arguments to the flagset passed in as an
+// argument.
+func (o *unjoinFederationOptions) Bind(flags *pflag.FlagSet) {
+	flags.StringVar(&o.clusterContext, "cluster-context", "",
+		"Name of the cluster's context in the local kubeconfig. Defaults to cluster name if unspecified.")
+	flags.BoolVar(&o.removeFromRegistry, "remove-from-registry", false,
+		"Remove the cluster from the cluster registry that is aggregated with the kubernetes API server running in the host cluster context.")
```

Comment:
we do use CRD for cluster registry too now. Need to update the info message.

Reply:
Done